### DISCUSSION TITLE
Unexpected behavior using libvips 8.13 applying filter to image and saving as AV1

### DISCUSSION THREAD

**Discussion by johntrunc** [libvips/libvips#3087] [Help wanted]
Hi,

I have found another unexpected behavior for libvips 8.13 compared to previous 8.12 version. 

For the following example I'm trying to thumbnail a jpeg and apply a filter with some alpha, after that saving the image as heif with AV1 compression. 

The result with version 8.12 is the expected one (some kind of sepia filtered image) but the same code working with version 8.13 is resulting in a white image:

* Result with libvips 8.12 (expected result): [out-8-12-2.avif.zip](https://github.com/libvips/libvips/files/9742307/out-8-12-2.avif.zip)
* Result with libvips 8.13 (unexpected white image result): [out-8-13-2.avif.zip](https://github.com/libvips/libvips/files/9742310/out-8-13-2.avif.zip)

You can download the used `example.jpeg` image [here](https://upload.wikimedia.org/wikipedia/commons/b/b7/The_sculptures_of_two_mythical_giant_demons%2C_Thotsakan_and_Sahatsadecha%2C_guarding_the_eastern_gate_of_the_main_chapel_of_Wat_Arun%2C_Bangkok.jpg) if needed. 

Executing the script on version 8.13 and replacing the `vips_heifsave` call with a `vips_jpegsave` is working fine so I guess is something specific of a side effect of one of the operations that I'm doing and the call to `vips_heifsave` function for libvips 8.13.

Not sure if this is a bug or not but it is working perfectly on version 8.12, so maybe something changed on version 8.13 that is affecting some of the operations done in the program and causing the unexpected final white image.

Thanks.

```c
// gcc -g -Wall testing-vips.c `pkg-config vips --cflags --libs`
#include <stdio.h>
#include <vips/vips.h>

void thumbnail_avif() {
  VipsImage *in, *out, *thumbnail, *filter, *filter_matrix, *filter_matrix_alpha, *matrix;

  double matrix_array[] = {
    0.3588, 0.7044, 0.1368,
    0.2990, 0.5870, 0.1140,
    0.2392, 0.4696, 0.0912
  };

  double alpha[] = {220.0};

  if (vips_jpegload("example.jpeg", &in, "shrink", 8, "autorotate", TRUE, NULL) == -1) {
    vips_error_exit(NULL);
  }

  if (vips_thumbnail_image(in, &thumbnail, 500, "height", 600, "size", VIPS_SIZE_BOTH, "crop", VIPS_INTERESTING_NONE, "no-rotate", TRUE, NULL) == -1) {
    vips_error_exit(NULL);
  }

  if (vips_copy(thumbnail, &filter, NULL) == -1) {
    vips_error_exit(NULL);
  }

  matrix = vips_image_matrix_from_array(3, 3, matrix_array, 9);
  if (matrix == NULL) {
    vips_error_exit(NULL);
  }

  if (vips_recomb(filter, &filter_matrix, matrix, NULL) == -1) {
    vips_error_exit(NULL);
  }

  if (vips_bandjoin_const(filter_matrix, &filter_matrix_alpha, alpha, 1, NULL) == -1) {
    vips_error_exit(NULL);
  }

  if (vips_composite2(thumbnail, filter_matrix_alpha, &out, VIPS_BLEND_MODE_OVER, "x", 0, "y", 0, NULL) == -1) {
    vips_error_exit(NULL);
  }

  if (vips_heifsave(out, "/tmp/out.avif", "Q", 75, "compression", VIPS_FOREIGN_HEIF_COMPRESSION_AV1, "strip", TRUE, NULL) == -1) {
    vips_error_exit(NULL);
  }

  g_object_unref(in);
  g_object_unref(out);
  g_object_unref(thumbnail);
  g_object_unref(filter);
  g_object_unref(filter_matrix);
  g_object_unref(filter_matrix_alpha);
  g_object_unref(matrix);
}

int main(int argc, char **argv)
{
  if (VIPS_INIT("testing-vips")) {
    vips_error_exit(NULL);
  }

  thumbnail_avif();

  return(0);
}
```

**Comment by lovell** [comment]
Did you see the new `bitdepth` parameter added in 8.13.0? Depending on the software you're using to view the output, you may want to set this to 8 (i.e. not high bit rate) to ensure compatibility.

https://www.libvips.org/API/current/VipsForeignSave.html#vips-heifsave

**Reply by johntrunc** [reply]
I have modified the call to `vips_heifsave` using `bitdepth` with value `8` and still generating an empty white avif image (using Google Chrome to see the result):

```c
// gcc -g -Wall testing-vips.c `pkg-config vips --cflags --libs`
#include <stdio.h>
#include <vips/vips.h>

void thumbnail_avif() {
  VipsImage *in, *out, *thumbnail, *filter, *filter_matrix, *filter_matrix_alpha, *matrix;

  double matrix_array[] = {
    0.3588, 0.7044, 0.1368,
    0.2990, 0.5870, 0.1140,
    0.2392, 0.4696, 0.0912
  };

  double alpha[] = {220.0};

  if (vips_jpegload("example.jpeg", &in, "shrink", 8, "autorotate", TRUE, NULL) == -1) {
    vips_error_exit(NULL);
  }

  if (vips_thumbnail_image(in, &thumbnail, 500, "height", 600, "size", VIPS_SIZE_BOTH, "crop", VIPS_INTERESTING_NONE, "no-rotate", TRUE, NULL) == -1) {
    vips_error_exit(NULL);
  }

  if (vips_copy(thumbnail, &filter, NULL) == -1) {
    vips_error_exit(NULL);
  }

  matrix = vips_image_matrix_from_array(3, 3, matrix_array, 9);
  if (matrix == NULL) {
    vips_error_exit(NULL);
  }

  if (vips_recomb(filter, &filter_matrix, matrix, NULL) == -1) {
    vips_error_exit(NULL);
  }

  if (vips_bandjoin_const(filter_matrix, &filter_matrix_alpha, alpha, 1, NULL) == -1) {
    vips_error_exit(NULL);
  }

  if (vips_composite2(thumbnail, filter_matrix_alpha, &out, VIPS_BLEND_MODE_OVER, "x", 0, "y", 0, NULL) == -1) {
    vips_error_exit(NULL);
  }

  if (vips_heifsave(out, "/tmp/out.avif", "bitdepth", 8, "Q", 75, "compression", VIPS_FOREIGN_HEIF_COMPRESSION_AV1, "strip", TRUE, NULL) == -1) {
    vips_error_exit(NULL);
  }

  g_object_unref(in);
  g_object_unref(out);
  g_object_unref(thumbnail);
  g_object_unref(filter);
  g_object_unref(filter_matrix);
  g_object_unref(filter_matrix_alpha);
  g_object_unref(matrix);
}

int main(int argc, char **argv)
{
  if (VIPS_INIT("testing-vips")) {
    vips_error_exit(NULL);
  }

  thumbnail_avif();

  return(0);
}
```

**Comment by jcupitt** [comment]
Hi, yes `heifsave` now supports HDR, so your float image is being saved as an HDR AVIF. Float images use 0 - 1 for black to white, so your 0 255 image is just white.

As a workaround, you can add this just after the `recomb`:

```C
{
        VipsImage *x;

        if (vips_cast_uchar(filter_matrix, &x, NULL))
            vips_error_exit(NULL);
        g_object_unref (filter_matrix);
        filter_matrix = x;
}
```

heifsave should probably not save all float images as HDR. Maybe we should only enable this for images tagged as scRGB (which do use 0-1 for black to white)?

**Reply by jfcalvo** [reply]
It is possible to define the `matrix_array` in a different way instead of doing the `vips_cast_uchar`? (I'm having a similar problem)

**Reply by jcupitt** [reply]
`recomb` always outputs float, so you need to cast after that if you want to be 8 bits.

Looking at the code I think this is an error in the new 16-bit HEIF path. I'll try to fix this for the next 8.13 release..

**Comment by jcupitt** [comment]
I improved the rules it uses for 16-bit heifsave: https://github.com/libvips/libvips/commit/fbef6746253bf26392eeb11b7e4106ac4ea3761a These now parallel the ones it uses for PNG save, so it should be more consistent.

This will be in 8.13.3. Thanks for reporting this @johntrunc !

**Reply by johntrunc** [reply]
Thanks to you @jcupitt 

Does than means that after the fix on 8.13.3 I don't need to do the `vips_cast_uchar` after the `recomb` call?

**Reply by johntrunc** [reply]
@jcupitt I have updated to 8.13.3 and now I can see the output AV1 image as expected (great!). The problem is that I can see some green artifacts so probably something is not working as expected:

[out.avif.zip](https://github.com/libvips/libvips/files/9946473/out.avif.zip)

Using my C example program and saving the ouput as PNG is returning an expected correct image without those green artifacts.

At this point I don't know if I'm not doing something right or it is your last change to heifsave. Could you please take a look to it? Thanks as always.

**Reply by jcupitt** [reply]
It's working for me, I ran:

```
$ gcc -g -Wall trunc.c `pkg-config vips --cflags --libs`
$ ./a.out ~/pics/example.jpg x.avif
$ vips copy x.avif x.png
```

To make:

![x](https://user-images.githubusercontent.com/580843/200265983-2501da45-b5d7-4c72-905d-91f356b4837d.png)

I tried with 8.13 and 8.14 (git master libvips).

That's with my version of your code, I'll try your original.

**Reply by jcupitt** [reply]
Yes, your code works too. I'm on ubuntu 22.04, with their standard libheif 1.12.0, libaom 3.3.0 packages.

**Reply by johntrunc** [reply]
My fault, I missed to add the call to `vips_cast_uchar` after vips_recomb` and now it's working fine.

One last question @jcupitt. Why is the `vips_cast_uchar` not necessary when saving the output as PNG? (But if missing when saving the output as AVIF is generating those green artifacts).

Thanks for taking the time to help me with this.

**Reply by jcupitt** [reply]
8.13 has different rules for AVIF -- it tries to save float images as HDR.

8.14 fixes this and PNG and AVIF use the same rules, so the extra cast won't be necessary.

**Reply by johntrunc** [reply]
I take note. Thank you very much @jcupitt and @lovell for your help.

**Comment by jcupitt** [comment]
You probably know, but your test prog could be simplified a little.

These lines:

```C
  if (vips_jpegload("example.jpeg", &in,
  if (vips_thumbnail_image(in, &thumbnail, 500,
```

Would be better as:

```C
  if (vips_thumbnail("example.jpeg", &thumbnail, 500, ...
```

You'll see better performance and lower memory use from doing the autorotate later.

You can skip the `copy` afterwards.

This line:

```C
  double matrix_array[] = {
    0.3588, 0.7044, 0.1368,
    0.2990, 0.5870, 0.1140,
    0.2392, 0.4696, 0.0912
  };
  matrix = vips_image_matrix_from_array(3, 3, matrix_array, 9);
```

Could be:

```C
  matrix = vips_image_new_matrixv(3, 3,  {
    0.3588, 0.7044, 0.1368,
    0.2990, 0.5870, 0.1140,
    0.2392, 0.4696, 0.0912
  });
```

Instead of:

```C
  double alpha[] = {220.0};
  if (vips_bandjoin_const(filter_matrix, &filter_matrix_alpha, alpha, 1, NULL)...
```

You could write:

```C
  if (vips_bandjoin_const1(filter_matrix, &filter_matrix_alpha, 220, NULL)...
```

**Reply by jcupitt** [reply]
I would also chain calls with `||`, and use a scope object. Perhaps:

```C
// gcc -g -Wall testing-vips.c `pkg-config vips --cflags --libs`
#include <stdio.h>
#include <vips/vips.h>

int thumbnail_avif(VipsObject *scope,
        const char *input_filename, const char *output_filename)
{
        VipsImage **t = (VipsImage **) vips_object_local_array(scope, 10);

        if (
                vips_thumbnail(input_filename, &t[0], 500,
                        "height", 600,
                        "size", VIPS_SIZE_BOTH,
                        "crop", VIPS_INTERESTING_NONE,
                        NULL) ||
                !(t[1] = vips_image_new_matrixv(3, 3,
                        0.3588, 0.7044, 0.1368,
                        0.2990, 0.5870, 0.1140,
                        0.2392, 0.4696, 0.0912)) ||
                vips_recomb(t[0], &t[2], t[1], NULL) ||
                vips_cast_uchar(t[2], &t[3], NULL) ||
                vips_bandjoin_const1(t[3], &t[4], 220, NULL) ||
                vips_composite2(t[0], t[4], &t[5], 
                        VIPS_BLEND_MODE_OVER, NULL) ||
                vips_heifsave(t[5], output_filename,
                        "bitdepth", 8,
                        "Q", 75, 
                        "compression", VIPS_FOREIGN_HEIF_COMPRESSION_AV1,
                        "strip", TRUE,
                        NULL))
                return -1;

        return 0;
}

int main(int argc, char **argv)
{
  VipsObject *scope;

  if (VIPS_INIT(argv[0]))
    vips_error_exit(NULL);

  scope = VIPS_OBJECT(vips_image_new());
  if (thumbnail_avif(scope, argv[1], argv[2])) {
          g_object_unref (scope);
          vips_error_exit (NULL);
  }
  g_object_unref (scope);

  return 0;
}
```

**Reply by johntrunc** [reply]
Thank you very much for the suggestions. I need to take a look to the documentation because the scope object as example is totally new to me!.
